diff --git a/ggml/src/ggml-metal/ggml-metal-impl.h b/ggml/src/ggml-metal/ggml-metal-impl.h
--- a/llama.cpp/ggml/src/ggml-metal/ggml-metal-impl.h
+++ b/llama.cpp/ggml/src/ggml-metal/ggml-metal-impl.h
@@ -8,6 +8,12 @@
 //
 // TODO: for optimal performance, become function of the device and work size
 
+#define N_R0_Q1_0 4
+#define N_SG_Q1_0 2
+
+#define N_R0_Q1_0_g128 4
+#define N_SG_Q1_0_g128 2
+
 #define N_R0_Q4_0 4
 #define N_SG_Q4_0 2
 
