--- llama.cpp/server/chat.mjs
+++ llama.cpp/server/chat.mjs
@@ -26,9 +26,8 @@ const propOrder = grammarJsonSchemaPropOrder

 let grammar = null
 if (grammarJsonSchemaFile) {
-    let schema = JSON.parse(readFileSync(grammarJsonSchemaFile, 'utf-8'))
-    const converter = new SchemaConverter({prop_order: propOrder, allow_fetch: true})
-    schema = await converter.resolveRefs(schema, grammarJsonSchemaFile)
+    const schema = JSON.parse(readFileSync(grammarJsonSchemaFile, 'utf-8'))
+    const converter = new SchemaConverter(propOrder)
     converter.visit(schema, '')
     grammar = converter.formatGrammar()
 }
diff --git llama.cpp/server/chat.sh llama.cpp/server/chat.sh
index da0a6ca..0143601 100755
